lisa/src/main/java/it/unive/lisa/analysis/nonrelational/Environment.java

```diff
@@ -155,6 +160,64 @@ public final boolean isBottom() {
 		return lattice.isBottom() && function == null;
 	}
 
+	@Override
+	public M pushScope(Call scope) throws SemanticException {
+		return this.liftIdentifiers(id -> new OutsideScopeIdentifier(id, scope));
+	}
+
+	@Override
+	public M popScope(Call scope) throws SemanticException {
+		AtomicReference<SemanticException> e = new AtomicReference<>();
+		M result = this.liftIdentifiers(id -> {
+			if (!(id instanceof OutsideScopeIdentifier)) {
+				return null;
+			} else {
+				Call otherCall = ((OutsideScopeIdentifier) id).getScope();
+				if (!scope.equals(otherCall) &&
+				// We might have a call that is resolved and the other one not,
+				// so we consider the program point as well
+						scope.getLine() != otherCall.getLine() && scope.getCol() != otherCall.getCol()) {
+					e.set(new SemanticException("Trying to pop out a different scope"));
+					return null;
+				} else {
+					try {
+						return ((OutsideScopeIdentifier) id).popScope(scope);
+					} catch (SemanticException semanticException) {
+						e.set(semanticException);
+						return null;
+					}
+				}
+			}
+		});
+		if (e.get() != null)
+			throw e.get();
+		else
+			return result;
+	}
+
+	private M liftIdentifiers(Function<Identifier, Identifier> lifter) throws SemanticException {
+		if (this.isBottom())
+			return this.bottom();
+		if (this.isTop())
+			return this.top();
+		Map<Identifier, T> function = new HashMap<>();
+		M result = this.copy();
+		for (Identifier id : this.getKeys()) {
+			Identifier outsideScopeIdentifier = lifter.apply(id);
+			if (outsideScopeIdentifier != null) {
+				T value = this.getState(id);
+				function.put(outsideScopeIdentifier, value);
```

Comment:
No: an environment could potentially apply additional reasoning when performing an assignment and we should always try to avoid propagating null values instead of creating new methods (you know where this leads). Create a method `mk()` with all the parameters you need